### PR TITLE
upload files into a collection starting from its show page

### PR DESCRIPTION
**Pull request by elrayle**
Steps to use this feature:
* On collection show page, a new action was added (under thumbnail) to 'Upload files' into the current collection being viewed

upload_files_from_collection

* Clicking 'Upload file' goes to the batch upload page and pre-selects the calling collection in the 'Upload to' selection menu.

batch_upload_to_collection

* Both 'Upload files' action on collection show page and 'Upload to' selection menu on batch upload page only appear when the upload_to_collection=true
* Text for actions on collection page and upload selection menu are configurable in sufia.en.yml

Changes also include...
* includes 2 fixes for rubocop unrelated to the changes for this feature

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -25,6 +25,7 @@
       post 'audit'
     end
   end
+  get 'files/new/in_collection/:collection_id' => 'generic_files#new', as: 'generic_files_upload_in_collection'
```
Comment (mjgiarlo) [inline config/routes.rb]: I'm probably overlooking something obvious, but why do we need a dedicated route for this rather than using the existing `/files/new` route with a param passed to it (e.g., `/files/new?collection=abc123def`)?

File: app/controllers/concerns/sufia/files_controller_behavior.rb [hunk 2]
```diff
@@ -64,6 +64,7 @@ def load_resource_from_solr
 
     # routed to /files/new
     def new
+      set_variables_for_new_form
```
Comment (mjgiarlo) [inline app/controllers/concerns/sufia/files_controller_behavior.rb]: Should we also be checking to make sure that current_user .can? edit collection? That'd be a good test to add.

File: spec/views/generic_file/new.html.erb_spec.rb [hunk 3]
```diff
@@ -16,7 +16,6 @@
       assign(:batch_id, batch_id)
       assign(:user_collections, user_collections)
       allow(controller).to receive(:current_user).and_return(stub_model(User))
-      allow(GenericFilesController).to receive(:upload_complete_path).with(batch_id).and_return("foo")
```
Comment (mjgiarlo) [inline spec/views/generic_file/new.html.erb_spec.rb]: Why was this line removed?

**Comment by mjgiarlo** [comment]
I have just a few questions.  Thanks for the PR, @elrayle

File: config/routes.rb (hunk 1 above)
Comment (elrayle) [inline config/routes.rb]: I'll look at making that adjustment.

File: app/controllers/concerns/sufia/files_controller_behavior.rb (hunk 2 above)
Comment (elrayle) [inline app/controllers/concerns/sufia/files_controller_behavior.rb]: The @user_collections which comes from hydra-collections returns only the list of collections the user has permissions controlled by line before_action on [line 43](https://github.com/projecthydra/sufia/blob/upload_with_collection/app/controllers/concerns/sufia/files_controller_behavior.rb) (find_collections_with_edit_access).  Then in set_variables_for_new_form, it checks that the passed in collection_id is in the @user_collections list.  Seems sufficient to me.  Does that work for you?

File: spec/views/generic_file/new.html.erb_spec.rb (hunk 3 above)
Comment (elrayle) [inline spec/views/generic_file/new.html.erb_spec.rb]: It didn't seem to be doing anything in the test since the test passes without it.  The only other test that uses something similar is controllers/generic_files_controller_spec.rb.  So honestly, I don't know why I put it in there in the first place.  It seems like kruft to me, so I removed it.

File: app/controllers/concerns/sufia/files_controller_behavior.rb (hunk 2 above)
Comment (elrayle) [inline app/controllers/concerns/sufia/files_controller_behavior.rb]: I'm adding two more tests that cover the scenario where a user tries to add a file to a collection they cannot edit.

Comment (mjgiarlo) [inline app/controllers/concerns/sufia/files_controller_behavior.rb]: :+1: Thanks much for indulging my paranoia. :)

File: config/routes.rb (hunk 1 above)
Comment (mjgiarlo) [inline config/routes.rb]: Thank you!

File: spec/views/generic_file/new.html.erb_spec.rb (hunk 3 above)
Comment (mjgiarlo) [inline spec/views/generic_file/new.html.erb_spec.rb]: :+1:

File: app/controllers/concerns/sufia/files_controller_behavior.rb (hunk 2 above)
Comment (elrayle) [inline app/controllers/concerns/sufia/files_controller_behavior.rb]: I never mind adding more tests.  I threw in a couple of others just to be more complete.

**Comment by elrayle** [comment]
Updated as recommended.

File: spec/controllers/generic_files_controller_spec.rb [hunk 4]
```diff
@@ -89,6 +103,16 @@
           expect(saved_file.to_solr['depositor_tesim']).to eq ['jilluser@example.com']
         end
 
+        it "adds new file when collection is is instructions id passed as numeric" do
+          xhr :post, :create, files: [file], Filename: "The world", batch_id: batch_id, permission: { "group" => { "public" => "read" } }, terms_of_service: "1", collection: -1
+          expect(response).to be_success
+        end
+
+        it "adds new file when collection is is instructions id passed as string" do
```
Comment (mjgiarlo) [inline spec/controllers/generic_files_controller_spec.rb]: Typo on lines 111 and 106: "is is".

File: sufia-models/app/actors/sufia/generic_file/actor.rb [hunk 5]
```diff
@@ -41,7 +41,7 @@ def create_content(file, file_name, path, mime_type, collection_id = nil)
     end
 
     def add_file_to_collection(collection_id)
-      return if collection_id.nil? || collection_id == -1
+      return if collection_id.nil? || ([-1, "-1"].include? collection_id)
```
Comment (mjgiarlo) [inline sufia-models/app/actors/sufia/generic_file/actor.rb]: Do you find that this ID is sometimes passed as a string and other times as a fixnum?

**Comment by elrayle** [comment]
@mjgiarlo, your paranoia is contagious.  I decided to write a test at the lowest level  to see if a user could add a file to a collection they do not own, and guess what, they can.  There are two places that the fix could go in, 1) collection.add_members, 2) a check before calling collection.add_members

I'm not sure hydra-collections knows anything about users, which would lean toward option #2.

I don't see a way that the user can pass a collection they don't own to [actor](https://github.com/projecthydra/sufia/blob/master/sufia-models/app/actors/sufia/generic_file/actor.rb).create_content (line 30) which calls add_file_to_collection (line 43) because there are several places in the UI that prevent that and users don't have a means to call that method directly.  But being real paranoid, if they some how did call actor.create_content with a collection they can't edit, the code will happily put the file in that collection.

Anyone want to weigh in on this?

**Comment by elrayle** [comment]
Ready for review.

**Comment by mjgiarlo** [comment]
Any thoughts on @elrayle's comment https://github.com/projecthydra/sufia/pull/1312#issuecomment-139546211 above, @cam156 @awead @jcoyne @hectorcorrea ?

**Comment by elrayle** [comment]
BTW, I took approach #2.

**Comment by mjgiarlo** [comment]
I'm inclined to :+1: but let's give the others a couple hours to take a look. @cam156 @jcoyne @hectorcorrea @awead

**Comment by awead** [comment]
I like it! One question: If the user can't edit the collection, maybe they shouldn't even be able to even attempt to add the files. It seems odd to me if they can go ahead and add the file, but then it doesn't get it added to the collection. Are there any messages informing them of this outcome?

**Comment by mjgiarlo** [comment]
@awead should a flash alert be populated there? would that suffice?

**Comment by awead** [comment]
@mjgiarlo sure, but I'm wondering, why not hide the "Add files" button entirely if they don't have edit access to the collection?

**Comment by elrayle** [comment]
@awead  ++  There is a check to see if the user can edit the collection in [_show_actions.html.erb](https://github.com/projecthydra/sufia/blob/upload_with_collection/app/views/collections/_show_actions.html.erb).  The actions (i.e., Edit, Add files, Upload files) only appear on the collection show if the user can edit.

**Comment by elrayle** [comment]
The test to be sure the user can edit before adding members is to stop malicious attacks.  If they go through the UI, they will never be offered the opportunity to add file to a collection they can't edit.

**Comment by awead** [comment]
ah, good then! thanks for the explanation, @elrayle

**Comment by cam156** [comment]
@elrayle have you done anything with the timing issue that occurs when more than one thread tries to update the collection?  We have has this issue in ArchiveSPhere, and @jcoyne just hit this issue in CurationConcerns with multi file upload.

**Comment by elrayle** [comment]
I was unaware of this issue.  Do you have an issue number for reference?

**Comment by elrayle** [comment]
@cam156 and others, would you be comfortable accepting this PR and opening an issue for the concurrent update problem?  This PR is not the cause of the concurrent update problem.  Additionally, this PR works in conjunction with PR https://github.com/projecthydra/sufia/pull/1307 which added the basic functionality on which this depends.  

Note that this functionality is turned off by default.

**Comment by mjgiarlo** [comment]
Sounds reasonable to me. Thanks, @elrayle!

**Comment by mjgiarlo** [comment]
Thanks too to @cam156 and @awead for comments.
